src/Lean/Widget/Diff.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2022 E.W.Ayers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: E.W.Ayers
+-/
+import Lean.Meta.PPGoal
+import Lean.Widget.InteractiveCode
+import Lean.Widget.InteractiveGoal
+import Lean.Data.Lsp.Extra
+
+namespace Lean.Widget
+
+open Server Std Lean SubExpr
+
+open Lean.Expr in
+/-- Define `m₁` to be the parent of `m₂` in a given mvar-context when `m₁` has been
+assigned with an expression containing `m₂`.
+
+Some technicalities:
+Suppose `m₁ : (n : Nat) → P`, and the new goal `n : Nat ⊢ m₂ : P` was got with an application of `intro`.
+Now, `m₁` is assigned the expression `fun (n : Nat) => m₃ n`.
+`m₃` is an auxilliary mvar with a delayed assignment, so we have to look a step further if there is a delayed mvar.
+
+-/
+partial def isParent (m₁ m₂ : MVarId): MetaM Bool := core m₁ 2
+  where core (m₁ : MVarId) (fuel : Nat) : MetaM Bool := do
+    let some assignment ← Lean.getExprMVarAssignment? m₁ | return false
+    let assignment ← instantiateMVars assignment
+    let r ← assignment.findExtM? (fun e => do
+      if !e.hasMVar then return FindStep.done else
+      match e with
+      | .mvar m =>
+        if m == m₂ then
+          return FindStep.found
+        let m' ← getDelayedMVarRoot m
+        if m' == m₂ then
+          return FindStep.found
+        if m' != m ∧ (← m'.isAssigned) ∧ fuel > 0 then
+          if ← core m' (fuel - 1) then -- [todo] is this wf without fuel?
```

Comment:
I don't really understand how the delayed mvars work in lean4. I was worried that there would be a way of getting an infinite descent with `getDelayedMVarRoot m` so I added a fuel param to the recursion.

The question is:

Let `m₁ ⊏ m₂` when `m₁ ≠ m₂` and `m₂ = getDelayedMVarRoot m₁` or `m₂` occurs in `m₁.assignment`. If I take the transitive closure of `⊏`, is it still irreflexive? It feels like it ought to be but I wasn't sure.

Reply:
I don't know whether searching through mvar assignments is the right thing to do here. Note that tactics could produce an arbitrarily long chain between an input and output mvar, but we definitely don't want to do an arbitrarily deep search through the proof term.

Reply:
Ok I've removed the recursion and it seems to work ok on my set of examples

Reply:
Ah no it doesn't work on `rw [h₁] at h₂`, you need to recurse to get the parent relationship. I am a bit worried about using a heuristic where you match up goals in zip order because it seems to me it's going to be easy to produce misleading output, particularly if the tactic is being applied to many goals with a combinator.

Reply:
I'll add a cache so at least it doesn't check the same mvar over and over.

Reply:
Fixed: I should have been using `Lean.Meta.getMVars`.